pallets/blindbox/src/lib.rs

```diff
@@ -0,0 +1,232 @@
+        BlindBoxesStillAvailable
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        #[pallet::weight(10_000)]
+        pub(super) fn set_blindbox_caller(origin: OriginFor<T>, account_id: T::AccountId ) -> DispatchResultWithPostInfo {
+            let _ = ensure_root(origin)?;
+
+            BlindBoxesCreators::<T>::insert(account_id, {});
+
+            Ok(().into())
+        }
+
+        #[pallet::weight(10_000)]
+        pub(super) fn generate_blindbox_ids(origin: OriginFor<T>, number_blindboxes: u32) -> DispatchResultWithPostInfo {
+            let caller = ensure_signed(origin)?;
+
+            // Ensure the authorized caller can call this func
+            ensure!(
+                BlindBoxesCreators::<T>::contains_key(&caller),
+                Error::<T>::NoPermission
+            );
+
+            // Ensure caller can only generate blindboxes once all the available blindboxes have been used
+            ensure!(
+                AvailableBlindBoxesCount::<T>::get() <= 0,
+                Error::<T>::BlindBoxesStillAvailable
+            );
+
+            let mut blindbox_vec = Vec::new();
+
+            // Generate random blindbox id and store
+            for x in 1..number_blindboxes +1 {
+                // Generate blindbox Id
+                let mut blindbox_id = Self::generate_random_number();
+
+                // Best effort attempt to remove bias from modulus operator.
+                for i in 1 .. T::MaxGenerateRandom::get() {
+                    blindbox_id = Self::generate_random_number();
+                }
+
+                // Push to Vec and save to storage
+                blindbox_vec.push(blindbox_id);
+                BlindBoxes::<T>::insert(blindbox_id, {});
+            }
+
+            AvailableBlindBoxesCount::<T>::put(number_blindboxes);
+
+            Self::deposit_event(Event::BlindBoxIdGenerated(blindbox_vec));
+
+            Ok(().into())
+        }
+
+        #[pallet::weight(10_000)]
+        pub(super) fn open_blind_box(origin: OriginFor<T>, blindbox_id: BlindBoxId) -> DispatchResultWithPostInfo {
+            let owner = ensure_signed(origin)?;
+
+            // Ensure the specified blindbox id exist in storage
+            ensure!(
+                BlindBoxes::<T>::contains_key(blindbox_id),
+                Error::<T>::BlindBoxDoesNotExist
+            );
+
+            let mut random_number = Self::generate_random_number();
+
+            // Best effort attempt to remove bias from modulus operator.
+            for i in 1 .. T::MaxGenerateRandom::get() {
+                random_number = Self::generate_random_number();
+            }
+
+            // 10% chance of winning
+            if random_number % 10 == 0 {
+                Self::save_blindbox_reward(&owner, blindbox_id);
+            }
+            // 50% chance of winning
+            else if random_number % 2 == 0 {
+                Self::save_blindbox_reward(&owner, blindbox_id);
+            }
+
```

Comment:
Where does the logic that you remove the current blindbox_id out of BlindBoxes? Otherwise, the current blindbox_id can open forever.